calico_node/startup.go

```diff
@@ -14,59 +14,657 @@
+	// set (which they might not otherwise be).
+	if err := ioutil.WriteFile("startup.env", []byte(text), 0666); err != nil {
+		log.WithError(err).Info("Unable to write to startup.env")
+		fatal("Unable to write to local filesystem")
+		terminate()
+	}
+}
+
+// getNode returns the current node configuration and a list of all configured
+// nodes.  If this node has not yet been created, it returns a blank node
+// resource.
+func getNode(client *client.Client, nodeName string) *api.Node {
+	meta := api.NodeMetadata{Name: nodeName}
+	node, err := client.Nodes().Get(meta)
+
 	if err != nil {
-		if _, ok := err.(errors.ErrorResourceAlreadyExists); !ok {
-			log.Warnf("Failed to set ClusterGUID: %s", err)
-			panic(err)
+		if _, ok := err.(errors.ErrorResourceDoesNotExist); !ok {
+			log.WithError(err).WithField("Name", nodeName).Info("Unable to query node configuration")
+			fatal("Unable to access datastore to query node configuration")
+			terminate()
 		}
-		log.Infof("Using previously configured ClusterGUID")
+
+		log.WithField("Name", nodeName).Info("Returning empty node configuration")
+		node = &api.Node{Metadata: api.NodeMetadata{Name: nodeName}}
+	}
+
+	return node
+}
+
+// configureIPsAndSubnets updates the supplied node resource with IP and Subnet
+// information to use for BGP.
+func configureIPsAndSubnets(node *api.Node) {
+	// If the node resource currently has no BGP configuration, add an empty
+	// set of configuration as it makes the processing below easier, and we
+	// must end up configuring some BGP fields before we complete.
+	if node.Spec.BGP == nil {
+		log.Info("Initialise BGP data")
+		node.Spec.BGP = &api.NodeBGPSpec{}
+	}
+
+	// Determine the autodetection type for IPv4 and IPv6.  Note that we
+	// only autodetect IPv4 when it has not been specified.  IPv6 must be
+	// explicitly requested using the "autodetect" value.
+	//
+	// If we aren't auto-detecting then we need to validate the configured
+	// value and possibly fix up missing subnet configuration.
+	ipv4Env := os.Getenv("IP")
+	if ipv4Env == "autodetect" || (ipv4Env == "" && node.Spec.BGP.IPv4Address == nil) {
+		adm := os.Getenv("IP_AUTODETECT_METHOD")
+		//node.Spec.BGP.IPv4Address, node.Spec.BGP.IPv4Network = autoDetectIPAndNetwork(adm, 4)
+		node.Spec.BGP.IPv4Address, _ = autoDetectIPAndNetwork(adm, 4)
+
+		// We must have an IPv4 address configured for BGP to run.
+		if node.Spec.BGP.IPv4Address == nil {
+			fatal("Couldn't autodetect a management IPv4 address:")
+			message("  -  provide an IP address by configuring one in the node resource, or")
+			message("  -  provide an IP address using the IP environment, or")
+			message("  -  if auto-detecting, use a different autodetection method.")
+			terminate()
+		}
+
+	} else {
+		node.Spec.BGP.IPv4Address, _ = fetchAndValidateIPAndNetwork(
+			ipv4Env, 4,
+			node.Spec.BGP.IPv4Address, nil)
+	}
+
+	ipv6Env := os.Getenv("IP6")
+	if ipv6Env == "autodetect" {
+		adm := os.Getenv("IP6_AUTODETECT_METHOD")
```

Comment:
I think we should not disallow that!  And if the method is "ping a particular address" rather than "ping" (with a separate address parm) then we deffo will.